### ISSUE TITLE
Data caching module a la `Sampler`

### ISSUE THREAD

**Issue by meder411** [pytorch/pytorch#37300]
## 🚀 Feature

A data caching module that can be instantiated as passed into a dataloader, similar to the way a sampler or dataset can currently. This module would accept a cache path, epoch to cache on, train/test flags if different behavior is desired depending on the mode, and perhaps other fields as well, depending on a set of most common needs. I imagine there would be some flag, `write_cache` that when `True`, would lead  `__call__()` to accept a list of corresponding data (e.g. `[img, labels]` or something like that) and cache it to the desired location. In `write_cache=False`, it reads from the the cache according to the dataloader, if the requested file exists in cache, and reads from the original location if the file is not in cache.

## Motivation

Caching can speed up training significantly for many datasets and networks. I am often finding that data loading speed is the biggest bottleneck to my training routines. When I write my own caching tool into the dataloader, I can sometimes see 10x+ faster training (of course it depends on the network, datasets, preprocessing, etc).

From what I gather from forums and simply looking at other people's code, many people write their own caching functions already. It would be great if there were a canonical one, tested and optimized for PyTorch dataloaders.

## Pitch

I would like to see a `Cacher` class developed that can be instantiated and extended similar to the `Sampler` classes and accepted by PyTorch dataloaders. This class would fit well in the existing PyTorch design principle of extensible abstraction: providing a high-level library, accessible even to novices, that is also malleable and easily extended by users with more advanced needs.

## Alternatives

The alternative really is for people to keep doing what they're doing. Folks who know how to do it probably are fine continuing doing so. However, this will help the new users and potentially save time for the power users.
